src/network/netdev/batadv.c

```diff
@@ -0,0 +1,199 @@
+/* SPDX-License-Identifier: LGPL-2.1+ */
+
+#include <inttypes.h>
+
+#include "netdev/batadv.h"
+#include "network-internal.h"
+#include "networkd-manager.h"
+#include "netlink-util.h"
+#include <linux/genetlink.h>
+
+#include "fileio.h"
+#include "parse-util.h"
+#include "stdio-util.h"
+#include "string-util.h"
+
+static void batadv_init(NetDev *n) {
+        BatmanAdvanced *b;
+
+        b = BATADV(n);
+
+        /* Set defaults */
+        b->aggregation            = true;
+        b->gateway_bandwidth_down = 10000;
+        b->gateway_bandwidth_up   = 2000;
+        b->bridge_loop_avoidance  = true;
+        b->distributed_arp_table  = true;
+        b->fragmentation          = true;
+        b->hop_penalty            = 15;
+        b->originator_interval    = 1000;
+}
+
+int config_parse_batadv_gateway_mode(
+                const char *unit,
+                const char *filename,
+                unsigned line,
+                const char *section,
+                unsigned section_line,
+                const char *lvalue,
+                int ltype,
+                const char *rvalue,
+                void *data,
+                void *userdata) {
+
+        BatmanAdvanced *b = userdata;
+        int mode;
+
+        assert(filename);
+        assert(lvalue);
+        assert(rvalue);
+        assert(data);
+
+        mode = BATADV_GW_MODE_OFF;
+
+        if (streq(rvalue, "client"))
+                mode = BATADV_GW_MODE_CLIENT;
+
+        else if (streq(rvalue, "server"))
+                mode = BATADV_GW_MODE_SERVER;
+
+        else {
+                log_syntax(unit, LOG_WARNING, filename, line, 0,
+                          "Batadv gateway mode '%s' unknown. Ignoring.", rvalue);
+                return 0;
+        }
+
+        b->gateway_mode = mode;
+
+        return 0;
+}
+
+int config_parse_badadv_bandwidth (
+                const char *unit,
+                const char *filename,
+                unsigned line,
+                const char *section,
+                unsigned section_line,
+                const char *lvalue,
+                int ltype,
+                const char *rvalue,
+                void *data,
+                void *userdata) {
+
+        uint64_t k;
+        uint32_t *bandwidth = data;
+        int r;
+
+        assert(filename);
+        assert(lvalue);
+        assert(rvalue);
+
+        r = parse_size(rvalue, 1000, &k);
+        if (r < 0) {
+                log_syntax(unit, LOG_WARNING, filename, line, r,
+                           "Failed to parse '%s=', ignoring assignment: %s",
+                           lvalue, rvalue);
+                return 0;
+        }
+
+        if (k/1000/100 > UINT32_MAX) {
+                log_syntax(unit, LOG_WARNING, filename, line, r,
+                           "The value of '%s=', is above UINT32_MAX: %s",
```

Comment:
UINT32_MAX is a C-ism. It's probably not too user-friendly to drop C-isms onto users, hence maybe say: "outside of 0…4294967295 range" or so.

BTW, what is the precise range accepted here? is it really the full 32bit range? the kernel doesn't refuse some of these values? for example 0, or enforces some other lower upper limit?

Reply:
I haven't checked the full source code of this PR but it looks like we are talking here about the announced gateway download/upload bandwidth. And the kernel accepts everything from 0 [100 kbit/s] to 4294967295 [100 kbit/s].